### PR TITLE
lib/errors: Add new classifiedError type

### LINKED ISSUE #38615: Create a custom error type to support error levels
The error type should:

1. Embed the original error and make it available to all consumers
2. Have two levels:
   - `warn`
   - `error` / `critical`

It remains to be decided which of the keywords, `error` or `critical` will be used to indicate the level. While using `error` allows for a direct one to one mapping of expected logging level, that is `log.Warn` for errors with `warn` level and `log.Error` for errors with `error` level. However, the word `error` itself makes for somewhat confusing code to read and write. For example:

```go
if err.Level == "error" {
    log.Error(err)
}
```

Versus

```go
if err.Level == "critical" {
    log.Error(err)
}
```

Having mentioned this, it remains a minor point and should not block the implementation of this issue. Choosing either is fine to start with and can be replaced easily if it feels like the other keyword is a better choice during the course of implementation of the overall Errors Classification project.

### PR DESCRIPTION
**Pull request by indradhanush**
Part of #38612.

Fixes #38615.

This error type is not used at the moment.

But it will be the next step.

## Test plan

New code that is not used anywhere yet. Also added some minimal tests to ensure this works for now. Will build upon it in future PRs.

### CODE REVIEW

File: lib/errors/levels.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+type ClassifiedError struct {
+	error error
+	level Level
+}
+
+func (ce *ClassifiedError) Error() string {
+	return ce.error.Error()
+}
+
+func (ce *ClassifiedError) IsLevelWarn() bool {
```
Comment (sashaostrikov) [inline lib/errors/levels.go:30]: nit: probably this method (and method below) can be shortened to `IsWarn` and `IsError`?

File: lib/errors/levels.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+func (ce *ClassifiedError) IsLevelError() bool {
+	return ce.level == LevelError
+}
+
+// Ensure that classifiedError always implements the error interface.
+var _ error = (*ClassifiedError)(nil)
+
+func NewClassifiedError(err error, l Level) error {
```
Comment (sashaostrikov) [inline lib/errors/levels.go:41]: probably it's worth moving this "constructor" function right under the type declaration, I guess it is pretty common to our codebase and also reads nicely when you know that there is a type and that's the function to create an instance of this type

File: lib/errors/level_test.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+		name     string
+		err      error
+		level    Level
+		expected string
+	}{
+		{
+			name:     "warn",
+			err:      New("warn error"),
+			level:    LevelWarn,
+			expected: "warn error",
```
Comment (sashaostrikov) [inline lib/errors/level_test.go:20]: this seems to be unused

**Review by sashaostrikov — APPROVED** [review]
LGTM overall, left a couple of comments

**Review by bobheadxi — COMMENTED** [review]
Requested changes:

1. We only export interface types in this package, and we should with this type as well IMO, mostly because of the next point:
2. It encourages smaller error types. For example in this case, what are we trying to solve? Do we want to _reduce_ the priority of certain errors? In that case, the _default_ case should be "an error is an error", but an _edge_ case could be "this is actually a _warning_ error". In this scenario, is it useful to say "an error is an error"? The better type to introduce would probably be `errors.Warning`, hypothetically:

```go
type Warning interface {
  error

  IsWarning()
}
```

3. Tests for errors should cover common use case paths (asserting the error)

**Pull request merged**